crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs

```diff
@@ -1289,6 +1313,103 @@ impl Drop for IndexeddbCryptoStore {
     }
 }
 
+/// Open the meta store.
+///
+/// The meta store contains details about the encryption of the main store.
+async fn open_meta_db(prefix: &str) -> Result<IdbDatabase, IndexeddbCryptoStoreError> {
+    let name = format!("{prefix:0}::matrix-sdk-crypto-meta");
+
+    debug!("IndexedDbCryptoStore: Opening meta-store {name}");
+    let mut db_req: OpenDbRequest = IdbDatabase::open_u32(&name, 1)?;
+    db_req.set_on_upgrade_needed(Some(|evt: &IdbVersionChangeEvent| -> Result<(), JsValue> {
+        let old_version = evt.old_version() as u32;
+        if old_version < 1 {
+            // migrating to version 1
+            let db = evt.db();
+
+            db.create_object_store("matrix-sdk-crypto")?;
+        }
+        Ok(())
+    }));
+
+    Ok(db_req.await?)
+}
+
+/// Load the serialised store cipher from the meta store.
+///
+/// # Arguments:
+///
+/// * `meta_db`: Connection to the meta store, as returned by [`open_meta_db`].
+///
+/// # Returns:
+///
+/// The serialised `StoreCipher` object.
+async fn load_store_cipher(
+    meta_db: &IdbDatabase,
+) -> Result<Option<Vec<u8>>, IndexeddbCryptoStoreError> {
+    let tx: IdbTransaction<'_> =
+        meta_db.transaction_on_one_with_mode("matrix-sdk-crypto", IdbTransactionMode::Readonly)?;
+    let ob = tx.object_store("matrix-sdk-crypto")?;
+
+    let store_cipher: Option<Vec<u8>> = ob
+        .get(&JsValue::from_str(keys::STORE_CIPHER))?
+        .await?
+        .map(|k| k.into_serde())
+        .transpose()?;
+    Ok(store_cipher)
+}
+
+/// Save the serialised store cipher to the meta store.
+///
+/// # Arguments:
+///
+/// * `meta_db`: Connection to the meta store, as returned by [`open_meta_db`].
+/// * `store_cipher`: The serialised `StoreCipher` object.
+async fn save_store_cipher(
+    db: &IdbDatabase,
+    export: &Vec<u8>,
+) -> Result<(), IndexeddbCryptoStoreError> {
+    let tx: IdbTransaction<'_> =
+        db.transaction_on_one_with_mode("matrix-sdk-crypto", IdbTransactionMode::Readwrite)?;
+    let ob = tx.object_store("matrix-sdk-crypto")?;
+
+    ob.put_key_val(&JsValue::from_str(keys::STORE_CIPHER), &JsValue::from_serde(&export)?)?;
+    tx.await.into_result()?;
+    Ok(())
+}
+
+/// Given a serialised store cipher, try importing with the given key.
+///
+/// This is a helper for [`IndexeddbCryptoStore::open_with_key`].
+async fn import_store_cipher_with_key(
+    key: &[u8; 32],
+    serialised_cipher: &[u8],
+    db: &IdbDatabase,
+) -> Result<StoreCipher, IndexeddbCryptoStoreError> {
+    let cipher = match StoreCipher::import_with_key(key, serialised_cipher) {
```

Comment:
I'm a bit uneasy about using the key directly here, not least because the same key (after an HKDF) is used to derive an AES key to encrypt things in the react sdk's store.

Should we add another pair of methods to `StoreCipher` which will use an HKDF on the provided key, rather than using it directly as a ChaCha20-Poly1305 key?

(Note that we can't do the HKDF at a higher level, because that would break the fallback to passphrase below.)

Reply:
> Should we add another pair of methods to StoreCipher which will use an HKDF on the provided key, rather than using it directly as a ChaCha20-Poly1305 key?

Hmm, can we do this here locally inside the IndexedDB store instead? This would avoid the need to expand the API surface of the `StoreCipher`.

Reply:
The thing is that, if we use an HKDF, really we would ideally use a salt, so we need somewhere to store the salt. We could store it separately of course, but it would go very nicely as another entry in the `KdfInfo` enum.

Reply:
> The thing is that, if we use an HKDF, really we would ideally use a salt

Yes, if you're not sure in the randomness of your input key material, but you said that you are using 256 bits of uniformly random data, just that it's base64 encoded.

Is that not the case?

Reply:
As used by the react-sdk, yes it's 256 bits of uniformly random data, and I don't think either an HKDF or a salt are particularly necessary. It's just that the same 256 bits are used to derive a separate key elsewhere; reusing the same cryptographic material for multiple keys is bad practice. And this is going to be annoying to change later.

I'm not that bothered. LMK

Reply:
Well yes, I'm not against the HKDF step, I'm questioning the necessity of the salt, which then requires you to store the salt, which requires to extend the `KdfInfo` enum.

You can provide the domain separation you're seeking by providing the `info` parameter to the HDKF function.

Reply:
ok, I'll do that.

Reply:
Now done.